Migration to Python3

TeddyAndrieux opened the pull request:
**Component**:

'salt', 'build'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

#2203 

**Summary**:

- Install salt-master and salt-minion Python3 versions
- Make salt unit tests working in Python3
- Bump salt to `3002.2`
- Add an "ugly" orchestrate to migrate Salt minion
- Remove non-breaking space characters from Solutions manifests

**Acceptance criteria**: 

- [x] Working boostrap
- [x] Working expansions
- [x] Working upgrade
- [x] Working downgrade
- [x] Working solution install
- [x] Working bootstrap restore
- [x] Working unit tests in Python3
---

Fixes: #2203

TeddyAndrieux committed:
salt: In downgrade orchestrate use `metalk8s.cmp_sorted`

With Python3 we can no longer use `cmp` in `sort` and `sorted` function
so use our salt module `metalk8s.cmp_sorted` in Downgrade orchestrate if
we are running with Python3, so that it works well with Python2 and
Pyton3

Refs: #2203

TeddyAndrieux committed:
tests,docs: Use `salt-ssh --raw-shell` command before expansions

In the past a `salt-ssh` command was needed prior to start the
expansions to accept the ssh key but it's not longer needed so remove
it.
Instead do a simple `salt-ssh --raw-shell` command to check that
salt-ssh work properly

TeddyAndrieux committed:
salt: Add a `states` to execute a function on a minion with raw ssh

Right now `salt.function` salt state module do not handle `roster` for
ssh and `raw_shell`, create a custom salt state module to run a
`saltutil.cmd` with all kwargs input so that we support `roster` and
`raw_shell`

Sees: https://github.com/saltstack/salt/issues/58662

TeddyAndrieux committed:
build,salt,script: Migrate to Salt Python3

Change salt-master image to use Python3 salt-master package and also
install Python3 dependencies, embed Python3 salt-minion version instead
of Python2.

In order to install Python3 salt-minion we need:
- to install python3
- to install python36-rpm as by default version comparaison for package
  installation is wrong
- to install python3 on nodes for being able to use `salt-ssh`

Sees: https://github.com/saltstack/salt/issues/58039
Sees: https://github.com/saltstack/salt/issues/57972
Fixes: #2203

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

slaperche-scality commented on tests/install/steps/test_expansion.py:
```diff
@@ -37,16 +37,12 @@ def declare_node(
 
 @when(parsers.parse('we deploy the node "{node_name}"'))
 def deploy_node(host, ssh_config, version, node_name):
-    accept_ssh_key = [
```
Seems unrelated to this PR.

And if this is really useless, then the doc needs an update as well: https://metal-k8s.readthedocs.io/en/development-2.6/installation/expansion.html#id3

So maybe this should be done in a dedicated PR (and keep this one only about py3 migration)

TeddyAndrieux replied:
Yes you right, for the doc I think it's still good to have it as a pre-check for expansion.

Edit: Oh it will not work anyway in the doc since we need python3 installed to have a working salt-ssh, ok I will update the doc

TeddyAndrieux replied:
I updated the documentation but I think I will keep it in this PR since it's changes needed for Python3 migration even if it's not directly related

gdemonet replied:
Maybe you could replace this command with the one suggested in docs (using `--raw-shell`)? I think we added it here to make sure the test would fail early if there was any issue with the key, and also because it's good practice to do in tests exactly what we have in docs.

TeddyAndrieux replied:
For me this one was here just because it was needed at the beginning to accept the ssh public key of the new node.
Anyway right I can add this check here like in the doc

slaperche-scality commented on salt/_modules/metalk8s.py:
```diff
@@ -5,6 +5,7 @@
 import logging
 import os.path
 import re
+import six
```
If we're moving to Python 3, why using the `six` compat layer?

Like:
> - replace `basestring` with `six.string_types`

Why not using `str` directly?

From https://docs.python.org/3.0/whatsnew/3.0.html:

> The builtin basestring abstract type was removed. Use str instead. The str and bytes types don’t have functionality enough in common to warrant a shared base class. The 2to3 tool (see below) replaces every occurrence of basestring with str.


---

Otherwise we're gonna have two migration PR?
2only -> 2-3 compat and then 2-3compat -> 3 only?

TeddyAndrieux replied:
It's more because for the moment we cannot rewrite the salt module unit test to work in python3, and we will do it in the future.
Also, I didn't tested yet but for upgrade purpose, we will maybe need to have a module compatible with both py2 and py3

TeddyAndrieux commented:
/help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` |

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

TeddyAndrieux commented:
/after_pull_request 2850

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/migrate-python-3` has diverged from
`development/2.7` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.7` into `improvement/migrate-python-3`
* Rebase `improvement/migrate-python-3` onto `origin/development/2.7`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

gdemonet commented on docs/installation/expansion.rst:
```diff
@@ -237,11 +237,25 @@ Open a terminal in the Salt Master container using
 #. Check that SSH access from the Salt Master to the new node is properly
    configured (see :ref:`Bootstrap SSH Provisioning`).
 
+   .. note::
+
+     You need to use `salt-ssh` as raw ssh since `salt-ssh` need Python3
+     installed to work properly (that will be installed by the following
+     salt orchestrate).
```
Not sure this note is relevant to our readers, but if it is, we should reformulate it a bit:

```
Salt SSH requires Python 3 to be installed on the remote host to run Salt functions.
It will be installed automatically when deploying the node, though you can send raw
shell commands before (using ``--raw-shell``) if needed.
```

(side-note: literal text needs double backticks **``** in RST (or the `:literal:` role), otherwise it's interpreted differently and we cannot guarantee how it will be displayed - see [the docs](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#interpreted-text))

TeddyAndrieux replied:
Right, 
I think it's relevant because they cannot run directly `test.ping` as it was documented before, I will change it as you suggest

gdemonet commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -6,13 +6,24 @@
     - tgt: {{ node_name }}
     - saltenv: metalk8s-{{ version }}
     - sls:
+      - metalk8s.salt.minion.dependencies
```
Shouldn't this be `include`d in `roles.minion` instead? I get that we need to run it first to avoid failing the installation of other packages, but can't we handle this ordering with `include`s?

TeddyAndrieux replied:
Nop we can't because if we start the salt process without this `python36-rpm` installed once it come to salt-minion installation it will not compare the package version as expected (even if python36-rpm get installed during this state) because salt internal module get loaded only at the beginning of the state (I think).

Edit: Hmm ok I just did a try and it seems to work when I include it directly in the roles.minion :thinking: don't really know why my last try failed, I will change it

gdemonet committed:
salt/unit-test: Migrate to Python 3

- removes `salttesting` dependency
- migrates away from deprecated `TestCase` methods
- migrates `Mock` objects call args checks

gdemonet committed:
salt: Prevent invalid escape sequence warnings

We need to pass `\;` to the `isoinfo` binary for reading file contents
from an ISO archive, but this escape sequence is not valid in Python,
and Python 3 warns about it.
We could have escaped the backslash, but instead rely on the raw string
notation for Python 3 (`r'my \escaped sequence'`).

TeddyAndrieux committed:
build,tests: Bump salt to `3002.2`

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/migrate-python-3` has diverged from
`development/2.7` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.7` into `improvement/migrate-python-3`
* Rebase `improvement/migrate-python-3` onto `origin/development/2.7`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

TeddyAndrieux committed:
salt: Add a `migrate_salt` state for Upgrade/Downgrade

Since we migrate from Python2 to Python3 a "classic" salt states cannot
handle it properly as the Python version change during the state
execution.
Add a dedicated orchestrate that handle the migration from Salt Python2
to Salt Python3 and also the migration from Salt Python3 to Salt
Python2, call this new orchestrate during Upgrade and Downgrade if
needed

Refs: #2203

TeddyAndrieux committed:
salt: Use `loop.previtem` in Upgrade and Downgrade orchestrate

Since we migrate to Salt Python3 we have Jinja 2.11 so we can
use `loop.previtem` instead of ugly hack using a jinja variable.
NOTE: This hack is still needed in downgrade orchestrate when we
downgrade to `2.6.x` as we are running with Salt master metalk8s-2.6.x
that use Salt Python2 and Jinja 2.7

Refs: #2203

TeddyAndrieux committed:
salt: Do not call etcd state if Salt upgrade/downgrade failed

In `deploy_node` orchestrate, before this commit, during upgrade and
downgrade the `Install etcd node` state is called even if the Salt
upgrade/downgrade failed.
This commit just add a require on this etcd state

TeddyAndrieux committed:
salt: Remove non-breaking space characters from Solutions

Solutions YAML files contains some non-breaking space characters inside
some Jinja formula `{{ }}` that make Salt Python3 rendering fail.
NOTE: Those non-breaking space characters did not break Jinja2.7 used by
Salt in Python2 that's why we didn't see any issue because of this
before

gdemonet commented on salt/metalk8s/salt/minion/dependencies.sls:
```diff
@@ -0,0 +1,17 @@
+# Salt need `python36-rpm` to do proper version comparaison and since we
+# use salt-ssh to install salt the first time, we want this rpm to be installed
+# before Salt-minion as salt-minion installation state will report a "failure"
+# because of version comparaison
```
minor: `s/comparaison/comparison/`

TeddyAndrieux committed:
changelog: Add entry for Python3 migration

Add an entry in changelog file about Salt migration to Python3 and bump
to Salt version 3002.2

Refs: #2203

gdemonet commented on salt/metalk8s/orchestrate/migrate_salt.sls:
```diff
@@ -0,0 +1,24 @@
+#  - for downgrade from 2.7.x to 2.6.x (migration from Python3 to Python2)
+#
+# Salt migration cannot be done automatically by a salt state because of the
+# changement between Python2 and Python3
```
minor: `s/changement/change/`

gdemonet reviewed (dismissed):
Only really minor comments, otherwise LGTM! Will manually test as well

gdemonet committed:
salt: Migrate sparse volume cleanup to Python3

See: #2726
See: #2936

TeddyAndrieux commented:
/approve

bert-e commented:
# Conflict with a changeset in the queue



The changeset in this pull request conflicts with another changeset
already in the queue. Please wait for the current queue to merge into
the development branch. The conflict will then appear in this pull
request and can be sorted on the feature branch directly.

This changeset has *not* been added to the queue. 



*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



Please check the status of the associated issue None.

Goodbye teddyandrieux.
